Multiple accounts

C0ntroller opened the pull request:
### Pull Request

#### Description
I proprose the following changes in my PR:
- Encryption level is level 3
  - data is stored in local storage field `udata`
  - data is stored in format `{'platform (eg zih)': {'user': 'encrypted', 'pass': 'encrypted'}, ...}`
  - older levels will be automatically converted to new format
- `set_user_data` command _can_ user platform argument (default is zih)
- same for `get_user_data`
- new command `check_user_data` to check if userdata is set without fetching it in "frontend"
- `asdf` and `fdsa` are renamed to `user` and `pass` everywhere in code

Referenced Issue: close #65

Referenced ToDo from [project-board](https://github.com/orgs/TUfast-TUD/projects/1): [ToDo card](https://github.com/orgs/TUfast-TUD/projects/1#card-69514313)

#### Type of change
- [ ] Bug fix (non-breaking change which fixes a bug)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that might cause existing functionality to not work as expected)

#### Further info
- [ ] This change requires a documentation update
- [ ] I updated the documentation accordingly, if required
- [x] I commented my code where its useful

#### Testing
We have 1500+ Users. Please test your changes thoroughly.
- [ ] Tested my changes on Firefox
- [x] Tested my changes on Chromium-Based-Browsers
- [x] Successfully run `npm run test` locally

#### Additional Information
Although it _should_ work fine, be aware that testing this branch should be done in a extra environment as an downgrade is not possible.

The HTML-Select in the settings page is currently kind of ugly. Someone could fix this or we just wait for the new design.

OliEfr commented:
I tested a bit and had errors. For example, I checked out develop-branch, saved my login-data, switched to feature/multiple_accounts and reloaded the extension. Obviously, in that case the encryption level should update automatically and the auto-login should work. However, this wasnt the case.

Second, when I updated the encryption level and visited the settings page, I see the following. The statements `Derzeit sind keine Daten für diese Platform gespeichert.` and `Du hast Daten hinterlegt und wirst überall wo möglich automatisch eingeloggt!` are obviously exclusive. I also dont think two status-messages are necessary.

I only have done some shallow testing, without going much into details.

![image](https://user-images.githubusercontent.com/31124624/139806678-d9000419-ea61-4118-beb8-a39a1faf6036.png)

OliEfr reviewed (changes requested):
See comment above

C0ntroller commented:
>  I also dont think two status-messages are necessary anyways.

I think they are. One for the currently selected platform, one for the 'isEnabled' switch.

OliEfr commented:
Understand.

From a users point of view it's too confusing right now (as it was for me).

I see the following options:

1) Leave the second status message. From my point of view one is sufficient. The user can always skip through all portals and check, whether data is stored.

2) adjust the second status message to sth like "Aktuell Daten für ZIH, Slub gespeichert." Also, I would make it visually clear that this status message is not directly related to the input formula. (E.g. vertical line or place it below or above the "Alle Daten löschen" Button).

C0ntroller commented:
> Intuitively I expect the following behavior:
> 
> * AutoLogin can be actived, as soon as one login is stored. If one login is stored, the switch gets true.
> * deactivating the AutoLogin Button deactivates for all platforms

This is exactly how it should work. In your last screenshot it only was bugged because converting the old credentials failed. 

From my POV multiple status messages for multiple different things are a good thing. Maybe we should address this in the redesign?

OliEfr commented:
Yes, this should be considered in the redesign.

However, we might have a release before redesign is implemented. That is why it should be implementent clearly for now, if not beautiful. So before merging I would still go for my suggestions 1) or 2) or another one, if you have one. The way it currently is, is not clear for the user.

C0ntroller commented:
I already kinda tested another login. 

Just add another `option`-Tag with a value (eg `slub`) and save something. Then you can take any existing contentScript (eg hisqis) and call `get_user_data` with the platform you defined and log the response. 

I would like to argue that this is sufficiant for this PR as a SLUB-autlogin is another topic for another PR.

OliEfr reviewed (approved):
Thanks, the status message is very good now in my eyes. Also very nice that you thought about changing the form content and usernameExtraHint.

I tested and think it works. However, as this is a major chage I'd like another one to test the PR **throughly**. @FugiMuffi could you do this? Then you can leave your review and we can go the 'official' way on this PR ;)
Oli

OliEfr commented on src/background.js:
```diff
@@ -552,45 +572,111 @@ async function getKeyBuffer () {
-async function userDataExists () {
-  const { asdf, fdsa } = await getUserData()
-  return asdf !== undefined && fdsa !== undefined
+async function userDataExists (platform = 'zih') {
```
Should this really default to `zih`all times?

OliEfr reviewed (changes requested):
Sorry. I found a bug.

Add a second login exactly like you mentions (add a case in settings.js and a option in settings.html).

Now, only save login-data for the second option, e.g. slub. Now, go to Dashboard and try to deactivate AutoLogin. It doesnt work. I think its because `check_user_data` always defaults to `zih`. I am not sure whether this should be the case. It seems odd, without diving deeper into the code.

Imagine, all your changes will be distributed to the users directly. For now, user feedback is only very good. Let's keep it that way. Small bugs will definetly decrease users satisfaction (and probably spam my e-mail inbox with improvement suggestions ;) )

I am not saying slub-login has to be implemented in this PR. However it would help a great deal to verify the functionality.

OliEfr commented:
I'll have a look later.

Out of interest @C0ntroller : Why exactly did you do this? Why did you force-push and what was the advantage of it?

```
C0ntroller force-pushed the feature/multiple_accounts branch from f95f481 to 47be66d 2 hours ago
```

C0ntroller commented:
> Out of interest @C0ntroller : Why exactly did you do this? Why did you force-push and what was the advantage of it?

I fixed it and then I fixed it again. But as I'm the only one working on this branch and didn't want to have a useless commit, I amended the second commit. But as the first was already pushed I just overwrote it by force pushing.

OliEfr commented:
Ah ok I understand, thanks

OliEfr commented:
When you save a login, the second status message should update automatically, right? This works if you save the slub-login. But if you save the ZIH/Selma-Login the second status message doesnt change.

C0ntroller commented:
> When you save a login, the second status message should update automatically, right? This works if you save the slub-login. But if you save the ZIH/Selma-Login the second status message doesnt change.

Should be fixed. sendMessage calls an async function and doesn't wait for it's return while other commands are executed. check_user_data was (_sometimes_) called before set_user_data

OliEfr commented:
Now, when I save Login-Data only for an additional platform (e.g. slub), it says `Derzeit gespeichert sind Logindaten für: ZIH/Selma, slub`, which should not be the case.

Each time I test I need to add that extra option-tag and the additional case.

Please test your changes before asking for review.

OliEfr commented:
Could you finish this by tomorrow so that I can write a mail to IT-security tomorrow evening?
Oli

OliEfr commented:
I am not sure why, but I wasnt ablt to reproduce this error today.

@FugiMuffi Could you please **thoroughly** test the changes proposed in this PR? It is about using Login-Data for multiple accounts. You can add addtional accounts by adding options in settings.html and an additional case in settings.js. Please also test all the edge cases you can think about.

OliEfr commented:
Please also test what happens, when the update is arriving for the user!

FugiMuffi reviewed (approved):
Everything seems to work fine. I tested (on FF and Chrome each) an upgrade with/without saved login, saving more logins, deleting logins, and saving only a login that is not the default option (zih).

To make this scale easier, the platform configs should be stored in a js object or even a json file instead of using a switch/case. That is an improvement that needs to be taken care of, but doesn't have to be right now.
